cbv3_django_prototype/cbv3_django_prototype/resources/serializers.py

```diff
@@ -1,15 +1,84 @@
 from rest_framework import serializers
 from .models import Resource
+from userauth.serializers import UserSerializer
 
 
-class ResourceSerializer(serializers.ModelSerializer):
+class TagsSerializerField(serializers.ModelField):
+
+    child = serializers.CharField()
+    default_error_messages = {
+            'not_a_list': ('Expected a list of tag names but got type "{input_type}".'),
+            'not_a_string': ('All list items must be type str.')
+                    }
+
+    def __init__(self, **kwargs):
+        super(TagsSerializerField, self).__init__(**kwargs)
+
+    def to_internal_value(self, value):
+        if not value:
+            value = []
+
+        if not isinstance(value, list):
+            self.fail('not_a_list', input_type=type(value).__name__)
+
+        for tag in value:
+            if not isinstance(tag, str):
+                self.fail('not_a_string')
+
+            self.child.run_validation(tag)
+
+        return value
+
+    def to_representation(self, instance):
+        tags = instance.tags.names()
+        slugs = instance.tags.slugs()
+
+        value = [{'slug': slug, 'name': name} for slug, name in zip(tags, slugs)]
```

Comment:
This does something very strange with two tags submitted for a resource when extended character sets are used. The `name` property for the first tag returned is set to `""` and for the second tag is set to `"_1"`. This behaviour seems to be consistent for a random selection of resources. The same issue appears if the resource is subsequently retrieved with the guid. It looks as if the `name` and `slug` fields have been swapped to how they appear in the database.

For example:

```json
"title": "Handmade Plastic Computer",
    "url": "https://larissa.com",
    "tags": [
        "relationships",
        "orchestrate",
        "initiative",
        "wireless monetize methodologies",
        "这是中文",
        "🤓"
    ]
}
```

```json
{
  "guid": "e6eb1a12-948d-4ed9-ab1f-a20eae291dee",
  "author": "",
  "title": "Handmade Plastic Computer",
  "description": "",
  "url": "https://larissa.com",
  "referring_url": "",
  "other_referring_source": "",
  "user": {
    "id": 1,
    "username": "Maximilian_Mante31",
    "first_name": "Monroe",
    "last_name": "Schuster",
    "email": "Katrina_Little@hotmail.com",
    "is_superuser": false
  },
  "date_published": "2020-02-18T12:09:36.685356-08:00",
  "created": "2020-02-18T12:09:36.685580-08:00",
  "modified": "2020-02-18T12:09:36.685364-08:00",
  "media_type": "",
  "paid": null,
  "tags": [
    {
      "slug": "initiative",
      "name": ""
    },
    {
      "slug": "orchestrate",
      "name": "_1"
    },
    {
      "slug": "relationships",
      "name": "initiative"
    },
    {
      "slug": "wireless monetize methodologies",
      "name": "orchestrate"
    },
    {
      "slug": "这是中文",
      "name": "relationships"
    },
    {
      "slug": "🤓",
      "name": "wireless-monetize-methodologies"
    }
  ]
}
```

Note how sorting appears to have been applied differently to `name` and `slug`.

Reply:
Congratulations!  You just found about 5 bugs in one shot!  Ever consider a career in QA??  (_Hee hee._)  

1.  I think we should log Unicode, extended and special character handling as an individual bug - we'll need to write out some specific use cases and test cases for those.

2.  We need to figure out if there are scenarios where we reject certain things out of hand for direct requests to the api.  Are there special characters or combinations we should look for/fence?  What does that look like?  The code checks for strings...but doesn't really have good fences, so that's its own issue.

3.  Would there be scenarios where we require the front end to sanitize or block certain entries?  If so, we should do some thinking about that an log some front-end issues accordingly.

4.  Do we pre-set or suggest allowable tags?  My first reaction is NO - that really limits the way in which users can interact with them.  HOWEVER, free-form tags set us up for all sorts of duplication and other processes, so I think we need some discussion around that.

5.  As for the **ordering** - you caught me!!  **P1** bug.  The comprehension reverses the order in the call to `zip()`, which is both insidious and unforgivable.  *bad developer!*  *bad!*  I will be fixing that in a push to this PR - because it will really screw with everyone otherwise.

Reply:
...as an additional note, I need a test case for that in the to-be-coming `tags` test cases!!

Reply:
As an additional note here:  If unicode or extended characters are used, it appears to break the "slugify" code, which then screws up the ordering before the serializer even gets its hands on it...and I am assuming it borks the write to the DB as well.  So we definitely need a bug for that, and a solution, which probably means modifying or replacing the code in the "slugify" module.

Reply:
Linked to #73 and also #72